feat: Add transform to strip duplicate spaces

richdouglasevans opened the pull request:
Closes #5679

This adds one of the text transforms -- strip duplicate spaces -- from issue #5659.

I'm using this PR to both flush out the approach and learn any things about contributing to this codebase. If all goes well I'll implement the other text transforms secure in the knowledge that the approach has been signed-off by Hendrik (and others) and any "dumb" first-time contributor stuff is sorted.

## Description

This adds a new _Transform_ submenu to the editor which has exactly one transform in it (for now): strip duplicate spaces.

In the video I am:

* selecting text that I want to strip duplicate spaces from
* clicking the command from the aforementioned submenu

After running the command the duplicate spaces are stripped: that is, collapsed to a single space.

https://github.com/user-attachments/assets/4ac026f3-52f4-44cc-b608-ebb48bfaa3cc

## Additional information

Tested on: Windows 11 Home.

> Note that I have not updated the CHANGELOG per the PR template guideline 'cos I don't know when this'll be merged, if ever 🤷

richdouglasevans commented on source/common/modules/markdown-editor/commands/transforms/transform-selected-text.ts:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @return  {StateCommand}            A `Command` to transform selected text.
+ */
+export function transformSelectedText (transform: TransformText): StateCommand {
```
This function encapsulates all the machinery of transforming selected text using the CodeMirror API.

Individual transforms are then plugged into this function -- via the `transform` parameter -- and don't need to duplicate logic and/or concern themselves with the CodeMirror API.

This is a `StateCommand` -- which just requires `state` and `dispatch` -- rather than a more general `Command` because it makes these transforms easier to (unit) test because we don't have to create a "full fat" `EditorView`.

* [StateCommand](https://codemirror.net/docs/ref/#state.StateCommand)

richdouglasevans commented:
I have another 7 transforms prepped and ready for PR-ing... I thought I'd get this one uop first because it demonstrates the approach and then, after making any changes from the reviewer, we can (hopefully) quickly get the rest of the transforms in.

nathanlesage commented:
I will need some more time to think about your style of coding because it is different from mine — not in a bad way, so this is not a criticism, but it's a bit more involved to think about this. A few things I already saw you may want to tackle in the meantime is:

1. I feel your unit tests are extremely repetitive. When you test for various inputs for a single function, for example, it is much simpler and easier to create an array of test cases at the top of the file, and implement the tester function once. This way the file is more organized, and it will be simple to just add additional test cases in the future.
2. I greatly appreciate your efforts to kickstart testing the codemirror related functionality. The main reason why we don't yet have such tests is literally just because I was too lazy to write those helper functions until now. What I would love to see is to have that a bit more generalized so that we can test more than just the transformation commands in the future. Please do not take this to completely over-engineer that; just start naming things "codemirror-test-utils" or something like that simply so that we know where to add additional functionality as the need arises and we start testing more things.
3. Please don't write longer strings by concatenating them, this looks horribly ugly 🙈 The 80 character column limit is not enforced in cases where the description just takes longer, that is fine. If you really do need multi-line strings (but not in the tests please), you can use backtick-strings.
4. I'll have to think if it's really adequate to have the transform commands implemented using a factory style, because it does add some complexity that is more difficult to disentangle in the future. My first intention was to literally just write command functions as-is (in the sense that they could be copied and pasted into different code bases and they work there, too). I do see the appeal of saving on some code, however, but I'm wondering what your thoughts are on my approach to the table editor commands (see PR #5243, specifically this folder: https://github.com/Zettlr/Zettlr/tree/table-editor/source/common/modules/markdown-editor/table-editor/commands)

I'll be back with a more thorough code review — also please feel free to remove the explanatory comments, because it makes focusing on the code more difficult than necessary. I appreciate the thoroughness, but it is easier to parse if you could transfer that information into your PR description and leave the files!

Thank you already for your help and work!

richdouglasevans commented:
> I feel your unit tests are extremely repetitive. When you test for various inputs for a single function, for example, it is much simpler and easier to create an array of test cases at the top of the file, and implement the tester function once. This way the file is more organized, and it will be simple to just add additional test cases in the future.

I have changed the test fixtures to conform to your preference.

* 9899d09cea50656a0c26a72318d8d6fd2375c3f7

> What I would love to see is to have that a bit more generalized so that we can test more than just the transformation commands in the future.

I have moved the supporting code into a folder per your sugggestion.

* 9128bda64435de3288e7f422ab8f0bd2e8872915

> Please don't write longer strings by concatenating them.

Updated to conform to your preference in 9899d09cea50656a0c26a72318d8d6fd2375c3f7

> remove the explanatory comments

I have deleted all of them.

> I'm wondering what your thoughts are on my approach to the table editor commands

Approach looks solid ✌️ In your case there are distinct differences between chunks of commands... while there is some repetition, it reads well enough.

I only refactored to the composition approach after writing the third text transform command because by then I noticed that I was repeating the same code in every command: in the specific case of these text transforms there's no difference to the CodeMirror algorithm piece... the composition allows the code that's distinct and relevant to just that transform to "pop." 

It's easy enough for me to un-compose it.

nathanlesage reviewed (approved):
Finally! Sorry that I wasn't able to check and review the PR before last week's conference, but I finally was able to sit down and take a look.

Here's how I'd like this to go: I'll merge your changes, so that they will be included in tomorrow's nightly and tell the users to test it out. Then any additional issues will resurface. A few things I noted (which can be fixed down the line):

* The `zapGremlins` might be turned into a RegExp with ranges instead of individual characters (makes it less verbose)
* The quote-fixing functions all assume en-US quotes, and would probably benefit from access to the quote config. I have a pretty extensive list of quotes [here](https://github.com/Zettlr/Zettlr/blob/develop/source/win-preferences/schema/autocorrect.ts). My personal hunch is that we should turn ALL types of magic quotes into straight quotes, but only turn straight single or double quotes to the configured type of quotes (do nothing if that option is set to straight quotes already)

But other than that I can literally see nothing else, so it should, as the pros say, just work™!

Thank you again!
